src/generic.jl

```diff
@@ -26,15 +26,19 @@ evaluate(dist::PreMetric, a, b) = dist(a, b)
 # Generic functions
 
 """
-    result_type(dist::PreMetric, Ta::Type, Tb::Type) -> T
-    result_type(dist::PreMetric, a::AbstractArray, b::AbstractArray) -> T
+    result_type(dist, Ta::Type, Tb::Type) -> T
+    result_type(dist, a::AbstractArray, b::AbstractArray) -> T
+    result_type(dist, a::Number, b::Number) -> T
 
 Infer the result type of metric `dist` with input type `Ta` and `Tb`, or input
 data `a` and `b`.
 """
-result_type(::PreMetric, ::Type, ::Type) = Float64 # fallback
-result_type(dist::PreMetric, a::AbstractArray, b::AbstractArray) = result_type(dist, eltype(a), eltype(b))
+result_type(::PreMetric, ::Type, ::Type) = Float64 # fallback in Distances
+result_type(f, a::Type, b::Type) = typeof(f(oneunit(a), oneunit(b))) # don't require `PreMetric` subtyping
```

Comment:
Are we sure we want to preserve units if inputs have them for all kinds of distances?

Reply:
I'm not sure; I guess that's a question for @johnnychen94, since the method came from #185. To me the "all kinds of distances" isn't super worrying since any distance could just provide a more specialized method, but to be honest I'm not super confident in the interplay between units and distances and maybe that could be tedious or painful.

Reply:
Yes, as we've discussed in the original issue #190, this change gives other packages the ability to reuse the name `result_type` without introducing type piracy. Because `Distances` defines `result_type`, the fallback should be defined in `Distances`.

Reply:
To be clear, `result_type` is exported by `Distances`.

https://github.com/JuliaStats/Distances.jl/blob/e50899d2f3814c5a5adae74d919782ff0d133405/src/Distances.jl#L13

Reply:
ah good point, it is exported!

I think @nalimilan's question here was about `oneunit`, any thoughts on that @johnnychen94?

> Are we sure we want to preserve units if inputs have them for all kinds of distances?

Reply:
I'm not sure if I understand the question clearly and I don't know what that question intends to suggest.

---

Firstly, I don't know why the fallback is `Float64` for `PreMetric`, it's probably written down when `result_type` wasn't exported and the main consideration is probably arithmetic performance inside Distances only.

FixedPointNumbers has introduced a better strategy `floattype(x)` to rewrite that `N0f8`( normalized `UInt8`) should get promoted to `Float32` instead of `Float16`. I think Distances should follow the same pattern and should not hard code the result type into `Float64`. People don't like changes, so it's okay to keep the internal `Distances` unchanged.

We're not changing `Distances` internal implementation here. As is said, we're adding a fallback that will be overridden by all the internal distances' implementation.

I think for any general function `f` (not limited to Distances itself), calling `f` on the units is a more generic fallback. Let me give two examples here:

For `Unitful`:

```julia
julia> using Unitful, Distances

julia> x = fill(999u"mm", 4, 4);

julia> y = fill(1u"m", 4, 4);

julia> cityblock(x, y)
2//125 m

julia> result_type(Cityblock(), x, y)
Quantity{Rational{Int64},𝐋,Unitful.FreeUnits{(m,),𝐋,nothing}}

julia> result_type(cityblock, x, y) # undefined but shouldn't be Float64
```

Also for JuliaImages, `ColorVectorSpaces` has defined a lot of basic operations for `Gray` and `RGB`, so it makes sense to let `result_type` be `Gray`/`RGB` if that's what they are supposed to be.

```julia
julia> using ImageCore, ColorVectorSpace

julia> x = rand(Gray{N0f8}, 4, 4);

julia> -(oneunit(eltype(x)), oneunit(eltype(x)))
Gray{N0f8}(0.0)
```

Reply:
OK. Actually I hadn't realized that `oneunit` is already used by existing methods:
https://github.com/JuliaStats/Distances.jl/blob/e50899d2f3814c5a5adae74d919782ff0d133405/src/metrics.jl#L216-L221

`Float64` is indeed not a great default. Do you know whether it's actually used in Distances ? Maybe we could remove the fallback method and always rely on `oneunit`.

Reply:
By removing this fallback, and observing the test errors, it tells me that only `BhattacharyyaDist`, `HellingerDist` and `Bregman` requires this fallback for its `colwise` implementation. Currently, it makes sense to me to remove this fallback (probably in another PR).

Reply:
Thanks for checking. I think it would be better to try removing the fallback in this PR. Otherwise the new method wouldn't actually be used for `PreMetric` so it would be misleading to call that `Distances.result_type`!